File: src/contexts/identity/domain/UserMail.ts
export class UserMail {
  private readonly emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

  constructor (private readonly mail: string) {
    if (!this.isValid(mail)) {
      throw new Error(`Invalid email format: ${mail}`)
    }
  }

  private isValid (mail: string): boolean {
    return this.emailRegex.test(mail)
  }

  public toString (): string {
    return this.mail
  }

  public equals (other: UserMail): boolean {
    return other.toString() === this.mail
  }
}

File: src/main/java/com_maboroshi/spring/contexts/identity/domain/UserMail.java
package com_maboroshi.spring.contexts.identity.domain;

import com_maboroshi.spring.contexts.identity.domain.errors.InvalidMailException;

import java.util.regex.Pattern;

public class UserMail {
  private final String mail;
  private static final Pattern EMAIL_REGEX = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");

  public UserMail(String mail) throws InvalidMailException {
    if (!isValid(mail)) {
      throw new InvalidMailException("Invalid mail value");
    }
    this.mail = mail;
  }

  private boolean isValid(String mail) {
    if (mail == null || !EMAIL_REGEX.matcher(mail).matches()) {
      return false;
    }
    return true;
  }

  public String toString() {
    return this.mail;
  }

  public boolean equals(UserMail newMail) {
    return this.mail.equals(newMail.toString());
  }
}
